src/number_theory/wilson.lean

```diff
@@ -0,0 +1,75 @@
+/-
+Copyright (c) 2022 John Nicol. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: John Nicol
+-/
+import number_theory.legendre_symbol.gauss_eisenstein_lemmas
+
+/-!
+# Wilson's theorem.
+
+This file contains a proof of Wilson's theorem.
+
+The heavy lifting is mostly done by the previous `wilsons_lemma`,
+but here we also prove the other logical direction.
+
+This could be generalized to similar results about finite abelian groups.
+
+## References
+
+* [Wilson's Theorem](https://en.wikipedia.org/wiki/Wilson%27s_theorem)
+
+-/
+
+open_locale nat
+
+namespace nat
+
+/-- For `n > 1`, `(n-1)!` is congruent to `-1` modulo `n` only if n is prime. --/
+private lemma wilsons_theorem_only_if_direction
+  (n : ℕ) (h : (((n - 1)! : zmod n) = -1)) (h1 : 1 < n) : (prime n) :=
+begin
+  have hp : ((n - 1)! + 1 : zmod n) = 0,
+  { rw h, simp, },
+  have hn_divides : n ∣(n-1)! + 1,
+  { rw ← zmod.nat_coe_zmod_eq_zero_iff_dvd, exact hp, },
+
+  by_contradiction h2,
+  obtain ⟨m⟩ := exists_dvd_of_not_prime2 h1 h2,
+  have m_leq_n_minus_one : m ≤ (n-1),
+  { cases h_1.right,
```

Comment:
```suggestion
  { cases h_1.right,
```
`h_1` (and `h_2` below) seems to be an auto-generated name and it's not clear where it comes from. In general, it's best to give it a name where it's introduced (the tactic that introduces it should allow names be supplied).